### ISSUE TITLE
mixed audio tracks sound fine on the timeline but are clipping in the export

### ISSUE THREAD

**Issue by PontiacCZ** [OpenShot/openshot-qt#3256]
**Describe the bug:**
When I add music to our birthday celebration, it all sounds good while played back on the timeline. But audible clipping is in the export file.

**Steps to reproduce the behavior:**
1. Download these two audio files:
    - [sound of people crowd](https://www.dropbox.com/s/mbzbnkfnd2y3qep/celebration_sounds.wav?dl=0) - 20 s
    - [Small guitar](https://www.bensound.com/royalty-free-music/track/small-guitar) music (free download)
2. import them into a new project in OpenShot 2.5.0
3. choose project profile: HD 1080p 25 fps
4. put celebration to track 4 and the music to track 5. It should look like this:
![image](https://user-images.githubusercontent.com/9153595/75376040-a2248b80-58cf-11ea-97ca-506a0a493e54.png)

5. play back first 20 secs of the timeline - you can hear that all sounds good in the OpenShot.
6. now export video: H265, profile HD 1080p 25 fps, quality "med", on the Advanced tab: start frame 1, end frame 500 (so you export only 20 seconds, we do not need more)
7. play back the exported file - you will hear pretty bad audio clipping, especially  with the drum beats. My encode is [here](https://www.dropbox.com/s/x1jwdflio9g943k/clipping.mp4?dl=0).

**Expected behavior:**
I expect the audio in the exported file to sound the same way as in the timeline. 

**System Details:**
- OpenShot Version: 2.5.0
- Operating System / Distro: Windows 10 Pro x64

I know there is already [an issue](https://github.com/OpenShot/openshot-qt/issues/1069) about this problem, but I wanted be more specific, in my report, provide steps and files for you to work with.

**Comment by zackees** [comment]
Yup - i have the same problem.

I'm using an MP3 audio track

**Comment by cleveopiyo** [comment]
Having the same problem with audio clipping but with mp4 audio and only after I export

**Comment by ferdnyc** [comment]
If you're mixing two audio sources together at full volume, clipping is a likely result. Especially if either or both of them are already close to or _at_ the limits of their amplitude range.

Loading the guitar track into Audacity, you can see that it's _already_ clipping some, and that's _**without**_ any other tracks mixed in:

(The red lines represent clipping, with View > Show Clipping enabled.)

![image](https://user-images.githubusercontent.com/538020/76161199-64add280-6107-11ea-9937-aeb96d7cf533.png)

So it's no surprise that when we mix in another fairly loud track at full volume, the clipping increases drastically:

![image](https://user-images.githubusercontent.com/538020/76161235-b48c9980-6107-11ea-8fd4-52fc1ff09e71.png)

And, yeah, it sounds pretty bad.

The trick is to simply reduce the volume of both tracks, so that they don't push each other beyond max amplitude when mixed together.

Here's the same mix, only this time I set Volume adjustment keyframes on the first frame of each clip (so the value applies for the entire length of the clip). I adjusted the guitar track to 0.75, and the crowd-noise track to 0.8:

![image](https://user-images.githubusercontent.com/538020/76161279-47c5cf00-6108-11ea-91dc-e224f61aedc6.png)

There's still a little _visible_ clipping, but just like with the original guitar track, none of it is audible.

The reason you don't hear clipping when _previewing_ is, you most likely have the output volume on your computer set far below 100%. So, there's plenty of headroom left in the output hardware, and the signal doesn't come anywhere near maximum amplitude — never mind exceeding it. If you were to play the preview back with the volume set to 100%, you'd notice that both tracks were far too loud, leading to distortion every time their combined amplitude attempted to exceed the maximum levels for your audio hardware

Long story short, when mixing audio from multiple sources, you have to reduce the input levels to prevent clipping. Don't assume that there _won't_ be clipping, and then plan to correct for it only if you hear distortion in the preview. Unless the track volumes are reduced, you **KNOW** there's going to be clipping, whether you hear it or not. So, just make the necessary adjustments right from the start. (And then plan to go back and tweak things some more, if the final result still doesn't sound exactly the way you planned.)

**Comment by SuslikV** [comment]
@ferdnyc the problem is slightly different, I think. At least, the last time I did my own test with the audio, the clipping itself wasn't an issue, but buffer data corruption caused major issue. The samples should be looked at high zoom before making conclusion. Clipping is just square signal that can cause distortion but buffer corruption is way different (random data). I want to say that your screenshots may not fully describe the original issue ( https://github.com/OpenShot/openshot-qt/issues/1069 ) that was mentioned in this thread.

**Comment by ferdnyc** [comment]
The other thing you can do, if you prefer not to deal with manually adjusting levels, is set the Volume Mixing mode for both (all) of your tracks to Average. In that mode, OpenShot will automatically average the volume of all playing tracks so that they don't exceed 1.0, which certainly works to prevent clipping. Here's the mix again, with both  tracks set to Volume Mixing: Average and Volume: 1.0... 

![image](https://user-images.githubusercontent.com/538020/76171635-d1ef5100-6163-11ea-87d7-8786ded6cc30.png)

Here's a breakdown of what's happening there:
* OpenShot automatically sets the volume of both tracks to 0.5 _while they are both playing_.
* Because audio volume is measured on a logarithmic scale, two tracks playing at 0.5 barely ever add up to more than 0.7 combined volume. (In _**theory**_ they could reach 1.0, but only if both tracks were at peak amplitude at _exactly_ the same time.)
* However, due to the logarithmic scale, a volume level of 0.5 is perceptually only a small reduction from 1.0 — it certainly doesn't sound "half as loud", in fact it's hardly noticeable.
* Nevertheless, when the crowd noise track ends, the guitar track does audibly increase in volume. (which at times can be a desirable effect, allowing background music to swell to the foreground when the dialogue and other environmental sounds mixed over it drop away.)

Note, BTW, that Volume Mixing only operates on the Volume _property_ for each track — it doesn't take into account the actual volume _levels_ of the audio. Any two tracks mixed with Volume Mixing: Average will both be set to 0.5 volume, even if the audio in one track is much louder than the other.

Volume Mixing _is_, however, compatible with other adjustments to the Volume property. If two tracks are both set to 0.8 Volume, Average mixing, then when they're combined they will each have a Volume level of 0.5. (**NOT 0.4.** Remember that Volume Mixing: Average only reduces the volume of the tracks enough to prevent going over 1.0 total.) When the second track ends, the first track will return to 0.8 volume instead of 1.0, which can make that transition less audible. Here's what that looks like, for our mix. 

(Upper waveform: 1.0,Average+1.0,Average; Lower waveform: 0.8,Average+0.8,Average):
![image](https://user-images.githubusercontent.com/538020/76172048-0bc25680-6168-11ea-9a24-76d48dba2f96.png)

As you can see, the levels are identical when both tracks are playing together, but the volume transition to the guitar track playing solo is less severe.

The reduction with Volume Mixing: Average may be a bit more severe than people would want, which is why it isn't enabled by default. It's _better_ to manually balance the audio levels where you want them, when mixing audio from multiple clips. But it's there as an option, if you prefer not to deal with levels. Another nice thing is that Volume Mixing: Average has no effect whatsoever when the clip is the only one playing audio, since there's nothing to average.

(There's a third Volume Mixing mode, Reduce, which simply sets the max volume of the clip to 0.77. That mode is stupid and nobody should use it.)

**Comment by ferdnyc** [comment]
@SuslikV 

Agreed, the audio mixing does appear to be doing some strange things when amplitude levels exceed 1.0, that makes the audio sound worse than normal clipping. That's a bug for libopenshot.

However, that corruption/distortion **only** happens when the audio levels exceed 1.0 (where there would otherwise be clipping anyway), so the practical solution to avoiding it is still just to _prevent_ the levels from exceeding max volume. As long as the amplitude levels stay safely below 1.0, the audio mixing is flawless.

(When they exceed it you get this — lower waveform — which I agree shouldn't happen.)

![image](https://user-images.githubusercontent.com/538020/76172244-285f8e00-616a-11ea-9429-879e66919be6.png)

**Comment by SuslikV** [comment]
>... (There's a third Volume Mixing mode, Reduce, which simply sets the max volume of the clip to 0.77. That mode is stupid and nobody should use it.)

The best comment I ever seen here. @ferdnyc , let's hope that there is serious math and experience behind this ^_^

**Comment by PontiacCZ** [comment]
> The reason you don't hear clipping when _previewing_ is, you most likely have the output volume on your computer set far below 100%. So, there's plenty of headroom left in the output hardware, and the signal doesn't come anywhere near maximum amplitude — never mind exceeding it. If you were to play the preview back with the volume set to 100%, you'd notice that both tracks were far too loud, leading to distortion every time their combined amplitude attempted to exceed the maximum levels for your audio hardware

You're right in this one.  My volumes are at about 30%:

![image](https://user-images.githubusercontent.com/9153595/76678673-f9b63d00-65d9-11ea-9e7b-ed22cd932de5.png)

So I did a test now: while playing back my timeline in OpenShot, I pulled the master volume up to nearly 90% (which was so loud that I started to worry about possible damage of my headphones and my ears but I did it only for a couple of seconds). No clipping appeared. 

![image](https://user-images.githubusercontent.com/9153595/76678964-a09bd880-65dc-11ea-88e1-d5016eb75afe.png)


On the contrary, I played back the exported file `clipping.mp4` on a very low volume so I could barely hear it. The clipping was still there.

I believe this is not about playback HW.

**Comment by SuslikV** [comment]
@ferdnyc Well, I thought that this is far worse error but it looks that it is just mistake in the code.
If you may notice how it works, just from the screenshots. What a beautiful math:

![OpenShot mixing overload errors 02](https://user-images.githubusercontent.com/19683044/76681939-5b899d80-6600-11ea-9b42-6f089c0981a9.png)


On the screenshot above is OpenShot's:
1) aac-compressed track (result of 1 stereo pre-overloaded track export)
2) aac-compressed 2 tracks mix (result of 2 stereo not overloaded tracks mix export) - bottom one.

The input samples in uncompressed `.wav` float format are simple as _Audacity_ main menu _Generate>DTMF Tones..._ (3 sec)
The overloaded sample in uncompressed `.wav` float format may be produced by _Audacity_ as own mix of two of these tones.

**Comment by weshinsley** [comment]
Two things I'd request as features...

1) Regardless of all the clipping issues above, it is really common for me that it sounds perfect when putting the movie together in Openshot, and the times it clips only come out in the final movie. That's hard to work with - I have to try and guess which channel is mainly responsible for the clipping at that moment, make the change in Openshot, re-export the video and see if I've got it. So, if audio when you're making the video could be the same as what gets exported, that would be amazing.

2) Rather than "Average" being applied per-channel, I'd like an overall "master" scaling done on the audio. I am usually mixing many videos together for a band - the largest number I've done is about 6 instruments and 12 voices. So 18 or so audio channels; averaging is all very well, but the final level of each track, and of the final output audio mix is extremely low. I have to pull the audio out in Audacity and normalise it up to a sensible level - but by that time I've lost quite a lot of signal.

Additionally, I find the audio much harder to mix in Openshot when the Average mode is enabled, since changes to one channel affect the others in the average... 

So the best way to solve this, I think, would be to have an option to calculate the maximum mixed audio level across the whole mixed audio, find the scaling factor to get that max down to 1 (or 0.99), and then multiply all the audio tracks levels by that scaling when mixing down, so the sum wouldn't clip.

**Comment by PontiacCZ** [comment]
Yep, I agree. I (as a user) do not need to understand the background or use some explicit type of mixing (e.g. average). I just expect that the sound in the output does not clip when it does not clip on the timeline. Common sense.
